### PR TITLE
KubeRay: kubebuilder creat RayService Controller and CR

### PR DESCRIPTION
**Pull request by brucez-anyscale**
## Why are these changes needed?
We plan to add a new RayService controller. This pr does the kubebuilder create api with basic implement. No real logic yet. Once this pr merged. We will start to implement logic for RayService.

## Related issue number
https://github.com/ray-project/kuberay/issues/214

## Checks

- [X] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
@Jeffwan @edoakes ping for review, thank you!

File: ray-operator/config/rbac/role.yaml
Comment (DmitriGekhtman) [inline ray-operator/config/rbac/role.yaml:115]: Hmm, just curious -- Why does the operator need permissions to create and delete RayServices?
(And why did the operator need permissions to create and edit RayClusters before the introduction of RayServices?)

Isn't the controller for a given resource is supposed to only read from spec and only update status?

File: ray-operator/config/rbac/role.yaml
Comment (DmitriGekhtman) [inline ray-operator/config/rbac/role.yaml:9]: Could you explain the changes to this file? Were these autogenerated?

File: ray-operator/config/crd/kustomizeconfig.yaml
Comment (DmitriGekhtman) [inline ray-operator/config/crd/kustomizeconfig.yaml:1]: What's this? Autogenerated?

File: ray-operator/config/crd/patches/cainjection_in_rayservice_rayservices.yaml
Comment (DmitriGekhtman) [inline ray-operator/config/crd/patches/cainjection_in_rayservice_rayservices.yaml:1]: Could you more about what this file is? Was it autogenerated?

File: ray-operator/config/crd/patches/webhook_in_rayservice_rayservices.yaml
Comment (DmitriGekhtman) [inline ray-operator/config/crd/patches/webhook_in_rayservice_rayservices.yaml:1]: Could you more about what this file is? Was it autogenerated?

File: ray-operator/Makefile
Comment (brucez-anyscale) [reply ray-operator/Makefile:43]: `manager-role` is auto generated by kubebuilder. I think previously, people just forget to change this to `kuberay-operator`

File: ray-operator/PROJECT
Comment (brucez-anyscale) [reply ray-operator/PROJECT:22]: I think we can use the same group. Usually the group name is the same as the generated package name. Then 2 controllers and CR will be in the same package folder.

File: ray-operator/apis/rayservice/v1alpha1/rayservice_types.go
Comment (brucez-anyscale) [reply ray-operator/apis/rayservice/v1alpha1/rayservice_types.go:8]: This one will auto generated if init by kubebuilder. But I do not think it is total empty

File: ray-operator/config/rbac/role.yaml
Comment (brucez-anyscale) [reply ray-operator/config/rbac/role.yaml:115]: I think every controller need permissions for its own CR. And RayService will maintain RayCluster, so I give RayService the RayCluster permission too.

File: ray-operator/config/rbac/role.yaml
Comment (brucez-anyscale) [reply ray-operator/config/rbac/role.yaml:9]: Auto generated and some manual fix to keep all existing rules.

File: ray-operator/config/crd/kustomizeconfig.yaml
Comment (brucez-anyscale) [reply ray-operator/config/crd/kustomizeconfig.yaml:1]: Autogenerated

File: ray-operator/config/crd/patches/cainjection_in_rayservice_rayservices.yaml
Comment (brucez-anyscale) [reply ray-operator/config/crd/patches/cainjection_in_rayservice_rayservices.yaml:1]: Autogenerated

File: ray-operator/config/crd/patches/webhook_in_rayservice_rayservices.yaml
Comment (brucez-anyscale) [reply ray-operator/config/crd/patches/webhook_in_rayservice_rayservices.yaml:1]: Autogenerated

File: ray-operator/config/rbac/rayservice_rayservice_editor_role.yaml
Comment (brucez-anyscale) [inline ray-operator/config/rbac/rayservice_rayservice_editor_role.yaml:1]: Autogenerated

File: ray-operator/config/rbac/rayservice_rayservice_viewer_role.yaml
Comment (brucez-anyscale) [inline ray-operator/config/rbac/rayservice_rayservice_viewer_role.yaml:1]: Autogenerated

File: ray-operator/config/samples/rayservice_v1alpha1_rayservice.yaml
Comment (brucez-anyscale) [inline ray-operator/config/samples/rayservice_v1alpha1_rayservice.yaml:1]: Autogenerated

File: ray-operator/config/rbac/role.yaml
Comment (DmitriGekhtman) [reply ray-operator/config/rbac/role.yaml:115]: Right -- the RayService controller needs to be able to create and delete RayClusters.

It's probably fine as is, but creating and deleting RayServices is not a function of the RayService controller -- users create/delete and the RayService controller responds. The RayService controller probably only needs write access to the status subresource.

File: ray-operator/PROJECT
Comment (DmitriGekhtman) [reply ray-operator/PROJECT:22]: The one practical consequence I can think of using one group is that all of the kinds in the group are versioned together. That actually makes sense given how closely related the resources are -- makes sense for v1alpha1  RayService to create v1alpha1 RayClusters.

It also matches how other projects do things --
For example OpenKruise has an [apps.kruise.io group](https://openkruise.io/docs/user-manuals/advancedstatefulset/) used by all of its workload CRs.

File: ray-operator/config/rbac/role.yaml
Comment (DmitriGekhtman) [reply ray-operator/config/rbac/role.yaml:115]: (And read access to the main resource.)

**Comment by Jeffwan** [comment]
Great! I will have a look later today.

**Review by DmitriGekhtman — APPROVED** [review]
Looks good to me.
Let's wait on review from @edoakes and @Jeffwan .

File: ray-operator/apis/ray/v1alpha1/rayservice_types.go
Comment (Jeffwan) [inline ray-operator/apis/ray/v1alpha1/rayservice_types.go:1]: I see you rename the fold and put everything under same `ray/v1alpha1`.  That means

1. All crds should have exact same group version
2. If we like to bump the version, we may bump all crd together? Or just create partial resources in v1alpha2 or v1beta1?

File: ray-operator/config/crd/patches/webhook_in_ray_rayservices.yaml
Comment (Jeffwan) [inline ray-operator/config/crd/patches/webhook_in_ray_rayservices.yaml:2]: at this moment, we don't use web hook. Instead of check in unuseful skeleton codes, I would suggest to remove them before check in. Could you share a list of the files (probably use tree) and we can discuss what folder we want to reserve?

File: ray-operator/config/rbac/role.yaml
Comment (Jeffwan) [reply ray-operator/config/rbac/role.yaml:9]: em.. chenk008 used to update the coordination for leadership election in https://github.com/ray-project/kuberay/pull/169/files. Seems the current script override the older fashion.  What's the kubebuilder version are you using?

Comment (brucez-anyscale) [reply ray-operator/config/rbac/role.yaml:9]: Default plugin keys: "go.kubebuilder.io/v3"
Default project version: "3"
controller-gen.kubebuilder.io/version: v0.6.0

I think I am using the same version

File: ray-operator/apis/ray/v1alpha1/rayservice_types.go
Comment (brucez-anyscale) [reply ray-operator/apis/ray/v1alpha1/rayservice_types.go:1]: @DmitriGekhtman I think we have a discussion before. We want to put them in the same group. But I am open to discuss that more.
From [kubebuilder](https://book.kubebuilder.io/multiversion-tutorial/tutorial.html), we should be able to maintain CRDs with different upgrading version. The crd types have versioning, but the controllers are the same. What required should be the conversion between versions.

**Comment by Jeffwan** [comment]
https://github.com/ray-project/kuberay/pull/275 is merged. Let's rebase the master and update this PR.

**Comment by brucez-anyscale** [comment]
> 

Thanks. I have rebased this pr.

File: ray-operator/config/rbac/role.yaml
Comment (brucez-anyscale) [reply ray-operator/config/rbac/role.yaml:115]: RayService and RayServices rabc are auto generated. I think RayService controller is the owner of RayService(s) CRD, so better to let it have all access.

Comment (DmitriGekhtman) [reply ray-operator/config/rbac/role.yaml:115]: Ok, sure.

**Comment by DmitriGekhtman** [comment]
Nice, the diff is much cleaner now.

**Review by Jeffwan — APPROVED** [review]
The changes look clean and straightforward.  
/lgtm

**Comment by Jeffwan** [comment]
Seems Dmitri has approved the changed as well and now we can merge the PR

**Pull request merged**
